File: udp-java/UDP/src/main/java/chat/dim/udp/Hub.java
/* license: https://mit-license.org
 *
 *  UDP: User Datagram Protocol
 *
 *                                Written in 2020 by Moky <albert.moky@gmail.com>
 *
 * ==============================================================================
 * The MIT License (MIT)
 *
 * Copyright (c) 2020 Albert Moky
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * ==============================================================================
 */
package chat.dim.udp;

import java.io.IOException;
import java.lang.ref.WeakReference;
import java.net.DatagramPacket;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.net.SocketException;
import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

/**
 *  Topology:
 *
 *                          +---------------+
 *                          |      APP      |
 *                          +---------------+
 *                              |       A
 *                              |       |  (filter)
 *                              V       |
 *          +-----------------------------------------------+
 *          |                                               |
 *          |     +----------+     HUB     +----------+     |
 *          |     |  socket  |             |  socket  |     |
 *          +-----+----------+-------------+----------+-----+
 *                   |    A                   |  |  A
 *                   |    |    (channels)     |  |  |
 *                   |    |                   |  |  |
 *          ~~~~~~~~~|~~~~|~~~~~~~~~~~~~~~~~~~|~~|~~|~~~~~~~~
 *          ~~~~~~~~~|~~~~|~~~~~~~~~~~~~~~~~~~|~~|~~|~~~~~~~~
 *                   |    |                   |  |  |
 *                   V    |                   V  V  |
 */

public class Hub extends Thread implements ConnectionHandler {

    private boolean running = false;

    // sockets
    private final Set<Socket> sockets = new LinkedHashSet<>();
    private final ReadWriteLock socketLock = new ReentrantReadWriteLock();

    // listeners
    private final Set<WeakReference<HubListener>> listeners = new LinkedHashSet<>();
    private final ReadWriteLock listenerLock = new ReentrantReadWriteLock();

    public Hub() {
        super();
    }

    //
    //  Listeners
    //

    public void addListener(HubListener listener) {
        Lock writeLock = listenerLock.writeLock();
        writeLock.lock();
        try {
            listeners.add(new WeakReference<>(listener));
        } finally {
            writeLock.unlock();
        }
    }

    public boolean removeListener(HubListener listener) {
        int count = 0;
        Lock writeLock = listenerLock.writeLock();
        writeLock.lock();
        try {
            Iterator<WeakReference<HubListener>> iterator = listeners.iterator();
            HubListener item;
            while (iterator.hasNext()) {
                item = iterator.next().get();
                if (listener.equals(item)) {
                    iterator.remove();
                    count += 1;
                    // break;
                }
            }
        } finally {
            writeLock.unlock();
        }
        return count > 0;
    }

    //
    //  Sockets
    //

    private Socket anySocket() {
        // Get arbitrary socket
        Socket socket = null;
        Lock readLock = socketLock.readLock();
        readLock.lock();
        try {
            Iterator<Socket> iterator = sockets.iterator();
            if (iterator.hasNext()) {
                socket = iterator.next();
            }
        } finally {
            readLock.unlock();
        }
        return socket;
    }

    private Set<Socket> allSockets() {
        // Get all sockets
        Set<Socket> all = new LinkedHashSet<>();
        Lock readLock = socketLock.readLock();
        readLock.lock();
        try {
            all.addAll(sockets);
        } finally {
            readLock.unlock();
        }
        return all;
    }

    private Set<Socket> getSockets(int port) {
        // Get all sockets bond to this port
        Set<Socket> matchedSockets = new LinkedHashSet<>();
        Lock readLock = socketLock.readLock();
        readLock.lock();
        try {
            InetSocketAddress address;
            for (Socket item : sockets) {
                if (port == item.port) {
                    matchedSockets.add(item);
                }
            }
        } finally {
            readLock.unlock();
        }
        return matchedSockets;
    }

    private Socket getSocket(int port) {
        // Get arbitrary socket bond to this port
        Socket socket = null;
        assert port != 0 : "port should not be ZERO";
        Lock readLock = socketLock.readLock();
        readLock.lock();
        try {
            InetSocketAddress address;
            for (Socket item : sockets) {
                if (port == item.port) {
                    socket = item;
                    break;
                }
            }
        } finally {
            readLock.unlock();
        }
        return socket;
    }

    private Socket getSocket(SocketAddress address) {
        // Get the socket bond to this address (host, port)
        Socket socket = null;
        Lock readLock = socketLock.readLock();
        readLock.lock();
        try {
            for (Socket item : sockets) {
                if (address.equals(item.localAddress)) {
                    socket = item;
                    break;
                }
            }
        } finally {
            readLock.unlock();
        }
        return socket;
    }

    protected Socket createSocket(SocketAddress address) throws SocketException {
        Socket socket = new Socket(address);
        socket.setHandler(this);
        new Thread(socket).start();
        return socket;
    }

    /**
     *  Create a socket on this address (port)
     *
     * @param address - address contains IP & port
     * @return socket
     * @throws SocketException on failed
     */
    public Socket open(SocketAddress address) throws SocketException {
        Socket socket;
        Lock writeLock = socketLock.writeLock();
        writeLock.lock();
        try {
            socket = getSocket(address);
            if (socket == null) {
                socket = createSocket(address);
                sockets.add(socket);
            }
        } finally {
            writeLock.unlock();
        }
        return socket;
    }

    public Socket open(String host, int port) throws SocketException {
        return open(new InetSocketAddress(host, port));
    }

    public Socket open(int port) throws SocketException {
        // get any socket bond to this port
        Socket socket;
        Lock writeLock = socketLock.writeLock();
        writeLock.lock();
        try {
            socket = getSocket(port);
            if (socket == null) {
                socket = createSocket(new InetSocketAddress(port));
                sockets.add(socket);
            }
        } finally {
            writeLock.unlock();
        }
        return socket;
    }

    /**
     *  Remove the socket on this address
     *
     * @param address - address contains IP & port
     * @return false on not found
     */
    public Set<Socket> close(SocketAddress address) {
        Set<Socket> closedSockets = new LinkedHashSet<>();
        Lock writeLock = socketLock.writeLock();
        writeLock.lock();
        try {
            Socket socket;
            while (true) {
                socket = getSocket(address);
                if (socket == null) {
                    break;
                }
                socket.close();
                closedSockets.add(socket);
                sockets.remove(socket);
            }
        } finally {
            writeLock.unlock();
        }
        return closedSockets;
    }

    /**
     *  Remove all sockets on this port
     *
     * @param port -
     * @return false on not found
     */
    public Set<Socket> close(int port) {
        Set<Socket> closedSockets = new LinkedHashSet<>();
        Lock writeLock = socketLock.writeLock();
        writeLock.lock();
        try {
            Socket socket;
            while (true) {
                socket = getSocket(port);
                if (socket == null) {
                    break;
                }
                socket.close();
                closedSockets.add(socket);
                sockets.remove(socket);
            }
        } finally {
            writeLock.unlock();
        }
        return closedSockets;
    }

    @Override
    public void start() {
        if (isAlive()) {
            return;
        }
        running = true;
        super.start();
    }

    /*
    public void stop() {
        // super.stop();
        close();
    }
     */

    public void close() {
        Lock writeLock = socketLock.writeLock();
        writeLock.lock();
        try {
            running = false;
            // close all sockets
            Iterator<Socket> iterator = sockets.iterator();
            Socket socket;
            while (iterator.hasNext()) {
                socket = iterator.next();
                socket.close();
                iterator.remove();
            }
        } finally {
            writeLock.unlock();
        }
    }

    //
    //  Connections
    //

    public Connection getConnection(SocketAddress destination, SocketAddress source) {
        Socket socket = getSocket(source);
        if (socket == null) {
            return null;
        }
        return socket.getConnection(destination);
    }

    private static Set<Connection> getConnections(SocketAddress destination, Set<Socket> sockets) {
        // Get connections from these sockets
        Set<Connection> connections = new LinkedHashSet<>();
        Connection conn;
        for (Socket sock : sockets) {
            conn = sock.getConnection(destination);
            if (conn != null) {
                connections.add(conn);
            }
        }
        return connections;
    }

    public Set<Connection> getConnections(SocketAddress destination, int source) {
        return getConnections(destination, getSockets(source));
    }

    public Set<Connection> getConnections(SocketAddress destination) {
        return getConnections(destination, allSockets());
    }

    /**
     *  Connect to the destination address by the socket bond to this source address
     *
     * @param destination - remote IP and port
     * @param source      - local IP and port
     * @return connection
     */
    public Connection connect(SocketAddress destination, SocketAddress source) {
        Socket socket = getSocket(source);
        if (socket == null) {
            return null;
        }
        return socket.connect(destination);
    }

    /**
     *  Connect to the destination address by any socket bond to this source port
     *
     * @param destination - remote IP and port
     * @param source      - local port
     * @return connection
     */
    public Connection connect(SocketAddress destination, int source) {
        Socket socket = getSocket(source);
        if (socket == null) {
            return null;
        }
        return socket.connect(destination);
    }

    /**
     *  Connect to the destination address by any socket exists
     *
     * @param destination - remote IP and port
     * @return connection
     */
    public Connection connect(SocketAddress destination) {
        Socket socket = anySocket();
        if (socket == null) {
            return null;
        }
        return socket.connect(destination);
    }

    private static Set<Connection> disconnect(SocketAddress destination, Set<Socket> sockets) {
        Set<Connection> allConnections = new LinkedHashSet<>();
        Set<Connection> removedConnections;
        for (Socket sock : sockets) {
            removedConnections = sock.disconnect(destination);
            if (removedConnections.size() > 0) {
                allConnections.addAll(removedConnections);
            }
        }
        return allConnections;
    }

    /**
     *  Disconnect from the destination address by the socket bond to this source address
     *
     * @param destination - remote IP and port
     * @param source      - local IP and port
     * @return removed connections
     */
    public Set<Connection> disconnect(SocketAddress destination, SocketAddress source) {
        Socket socket = getSocket(source);
        if (socket == null) {
            return null;
        }
        return socket.disconnect(destination);
    }

    /**
     *  Disconnect from the destination address by all sockets bond to this source port
     *
     * @param destination - remote IP and port
     * @param source      - local port
     * @return removed connections
     */
    public Set<Connection> disconnect(SocketAddress destination, int source) {
        return disconnect(destination, getSockets(source));
    }

    /**
     *  Disconnect from the destination address by all sockets
     *
     * @param destination - remote IP and port
     * @return removed connections
     */
    public Set<Connection> disconnect(SocketAddress destination) {
        return disconnect(destination, allSockets());
    }

    //
    //  Send / Receive
    //

    /**
     *  Send data to the destination address by the socket bound to this source address
     *
     * @param data        - UDP data package
     * @param destination - remote IP and port
     * @param source      - local IP and port
     * @return how many bytes have been sent
     */
    public int send(byte[] data, SocketAddress destination, SocketAddress source) {
        Socket socket = getSocket(source);
        if (socket == null) {
            return -1;
        }
        return socket.send(data, destination);
    }

    /**
     *  Send data to the destination address by any socket bound to this source port
     *
     * @param data        - UDP data package
     * @param destination - remote IP and port
     * @param source      - local port
     * @return how many bytes have been sent
     */
    public int send(byte[] data, SocketAddress destination, int source) {
        Socket socket = getSocket(source);
        if (socket == null) {
            return -1;
        }
        return socket.send(data, destination);
    }

    /**
     *  Send data to the destination address by any socket exists
     *
     * @param data        - UDP data package
     * @param destination - remote IP and port
     * @return how many bytes have been sent
     */
    public int send(byte[] data, SocketAddress destination) {
        Socket socket = anySocket();
        if (socket == null) {
            return -1;
        }
        return socket.send(data, destination);
    }

    private static void _sleep(double seconds) {
        try {
            sleep((long) (seconds * 1000));
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    private static Cargo receivePacket(Set<Socket> sockets) {
        // Receive data from these given sockets
        Cargo cargo = null;
        DatagramPacket packet;
        for (Socket sock : sockets) {
            packet = sock.receive();
            if (packet != null) {
                // got one
                cargo = new Cargo(packet, sock);
                break;
            }
        }
        return cargo;
    }

    private static Cargo receivePacket(Set<Socket> sockets, float timeout) {
        // Block to receive data from these given sockets
        long now = (new Date()).getTime(); // milliseconds
        long expired = now + (long) (timeout * 1000);

        Cargo cargo = null;
        while (now <= expired) {
            cargo = receivePacket(sockets);
            if (cargo != null) {
                // got it
                break;
            }
            // receive nothing, have a rest for next job
            _sleep(0.1);
            now = (new Date()).getTime();
        }
        return cargo;
    }

    /**
     *  Block to receive data
     *
     * @param source  - local IP and port
     * @param timeout - timeout in seconds
     * @return cargo with data and addresses
     */
    public Cargo receive(SocketAddress source, float timeout) throws IOException {
        Socket sock = getSocket(source);
        if (sock == null) {
            throw new IOException("socket not found: " + source);
        }
        Set<Socket> sockets = new LinkedHashSet<>();
        sockets.add(sock);
        if (timeout < 0) {
            timeout = FOREVER;
        }
        return receivePacket(sockets, timeout);
    }

    public Cargo receive(int port, float timeout) {
        Set<Socket> sockets = getSockets(port);
        if (timeout < 0) {
            timeout = FOREVER;
        }
        return receivePacket(sockets, timeout);
    }

    public Cargo receive(float timeout) {
        Set<Socket> sockets = allSockets();
        if (timeout < 0) {
            timeout = FOREVER;
        }
        return receivePacket(sockets, timeout);
    }

    private static final long FOREVER = 31558150; // 3600 * 24 * 365.25636 (365d 6h 9m 10s)

    public Cargo receive(SocketAddress source) throws IOException {
        Socket sock = getSocket(source);
        if (sock == null) {
            throw new IOException("socket not found: " + source);
        }
        Set<Socket> sockets = new LinkedHashSet<>();
        sockets.add(sock);
        return receivePacket(sockets);
    }

    public Cargo receive(int port) throws IOException {
        Socket sock = getSocket(port);
        if (sock == null) {
            throw new IOException("socket not found: " + port);
        }
        Set<Socket> sockets = new LinkedHashSet<>();
        sockets.add(sock);
        return receivePacket(sockets);
    }

    public Cargo receive() throws IOException {
        Set<Socket> sockets = allSockets();
        if (sockets.size() == 0) {
            throw new IOException("socket not found");
        }
        return receivePacket(sockets);
    }

    @Override
    public void run() {
        long now = (new Date()).getTime();
        long expired = now + Connection.EXPIRES;
        Cargo cargo;
        List<byte[]> responses;

        while (running) {
            try {
                // try to receive data
                cargo = receivePacket(allSockets());
                if (cargo == null) {
                    // received nothing, have a rest
                    _sleep(0.1);
                } else {
                    // dispatch data and got responses
                    responses = dispatch(cargo.data, cargo.source, cargo.destination);
                    for (byte[] res : responses) {
                        send(res, cargo.source);
                    }
                }
                // check time for next heartbeat
                now = (new Date()).getTime();
                if (now > expired) {
                    heartbeat();
                    // try heartbeat for all connections in all sockets
                    expired = now + 2000;
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    protected void heartbeat() {
        Lock readLock = socketLock.readLock();
        readLock.lock();
        try {
            for (Socket socket : sockets) {
                socket.ping();  // try to keep all connections alive
                socket.purge(); // remove error connections
            }
        } finally {
            readLock.unlock();
        }
    }

    private List<byte[]> dispatch(byte[] data, SocketAddress source, SocketAddress destination) {
        List<byte[]> responses = new ArrayList<>();
        Lock readLock = listenerLock.readLock();
        readLock.lock();
        try {
            HubListener listener;
            HubFilter filter;
            byte[] res;
            for (WeakReference<HubListener> item : listeners) {
                listener = item.get();
                if (listener == null) {
                    // TODO: remove empty item
                    continue;
                }
                filter = listener.getFilter();
                if (filter != null && !filter.checkData(data, source, destination)) {
                    continue;
                }
                res = listener.onDataReceived(data, source, destination);
                if (res == null) {
                    continue;
                }
                responses.add(res);
            }
        } finally {
            readLock.unlock();
        }
        return responses;
    }

    //
    //  ConnectionHandler
    //

    @Override
    public void onConnectionStatusChanged(Connection connection, ConnectionStatus oldStatus, ConnectionStatus newStatus) {
        Lock readLock = listenerLock.readLock();
        readLock.lock();
        try {
            HubListener listener;
            HubFilter filter;
            for (WeakReference<HubListener> item : listeners) {
                listener = item.get();
                if (listener == null) {
                    // TODO: remove empty item
                    continue;
                }
                filter = listener.getFilter();
                if (filter != null && !filter.checkConnection(connection)) {
                    continue;
                }
                listener.onStatusChanged(connection, oldStatus, newStatus);
            }
        } finally {
            readLock.unlock();
        }
    }

    @Override
    public void onConnectionReceivedData(Connection connection) {
        /*
        if (running) {
            // process by run()
            return;
        }
        Cargo cargo = null;
        try {
            cargo = receive(connection.localAddress);
        } catch (IOException e) {
            e.printStackTrace();
        }
        if (cargo == null) {
            // assert false : "receive error";
            return;
        }
        // dispatch data and got responses
        List<byte[]> responses = dispatch(cargo.data, cargo.source, cargo.destination);
        for (byte[] res : responses) {
            send(res, cargo.source);
        }
         */
    }
}

File: tcp-py/tcp/hub.py
# -*- coding: utf-8 -*-
#
#   TCP: Transmission Control Protocol
#
#                                Written in 2021 by Moky <albert.moky@gmail.com>
#
# ==============================================================================
# MIT License
#
# Copyright (c) 2021 Albert Moky
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
# ==============================================================================

import socket
import threading
import time
from abc import ABC
from typing import Optional, Iterable

from startrek.types import SocketAddress, AddressPairMap
from startrek.skywalker import Runnable, Runner, Daemon
from startrek import Channel, BaseChannel
from startrek import Connection, ConnectionDelegate
from startrek import BaseConnection, ActiveConnection
from startrek import BaseHub

from .aio import is_blocking
from .channel import StreamChannel


class ChannelPool(AddressPairMap[Channel]):

    # Override
    def set(self, item: Optional[Channel],
            remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        # 1. remove cached item
        cached = super().remove(item=item, remote=remote, local=local)
        if cached is not None and cached is not item:
            Runner.async_task(coro=cached.close())
        # 2. set new item
        old = super().set(item=item, remote=remote, local=local)
        assert old is None, 'should not happen: %s' % old
        return cached

    # Override
    def remove(self, item: Optional[Channel],
               remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        cached = super().remove(item=item, remote=remote, local=local)
        if cached is not None and cached is not item:
            Runner.async_task(coro=cached.close())
        if item is not None:
            Runner.async_task(coro=item.close())
        return cached


# noinspection PyAbstractClass
class StreamHub(BaseHub, ABC):
    """ Base Stream Hub """

    def __init__(self, delegate: ConnectionDelegate):
        super().__init__(delegate=delegate)
        self.__channel_pool = self._create_channel_pool()

    # noinspection PyMethodMayBeStatic
    def _create_channel_pool(self):
        return ChannelPool()

    #
    #   Channel
    #

    # noinspection PyMethodMayBeStatic
    def _create_channel(self, remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Channel:
        """ create channel with socket & addresses """
        return StreamChannel(remote=remote, local=local)

    # Override
    def _all_channels(self) -> Iterable[Channel]:
        """ get a copy of all channels """
        return self.__channel_pool.items

    # Override
    def _remove_channel(self, channel: Optional[Channel],
                        remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        """ remove cached channel """
        return self.__channel_pool.remove(item=channel, remote=remote, local=local)

    def _get_channel(self, remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        """ get cached channel """
        return self.__channel_pool.get(remote=remote, local=local)

    def _set_channel(self, channel: Channel,
                     remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        """ cache channel """
        return self.__channel_pool.set(item=channel, remote=remote, local=local)


class ServerHub(StreamHub, Runnable):
    """ Stream Server Hub """

    def __init__(self, delegate: ConnectionDelegate):
        super().__init__(delegate=delegate)
        self.__local = None   # SocketAddress
        self.__master = None  # socket.socket
        self.__daemon = Daemon(target=self)
        self.__running = False

    # Override
    def _create_connection(self, remote: SocketAddress, local: Optional[SocketAddress]) -> Optional[Connection]:
        conn = BaseConnection(remote=remote, local=local)
        conn.delegate = self.delegate  # gate
        return conn

    async def bind(self, address: Optional[SocketAddress] = None, host: Optional[str] = None, port: Optional[int] = 0):
        if address is None:
            if port > 0:
                assert host is not None, 'host should not be empty'
                address = (host, port)
            else:
                address = self.__local
                assert address is not None, 'local address not set'
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
        sock.setblocking(True)
        sock.bind(address)
        sock.listen(1)
        # sock.setblocking(False)
        self._set_master(master=sock)
        self.__local = address

    def _get_master(self) -> Optional[socket.socket]:
        return self.__master

    def _set_master(self, master: socket.socket):
        # 1. replace with new socket
        old = self.__master
        self.__master = master
        # 2. close old socket
        if not (old is None or old is master):
            old.close()

    @property
    def local_address(self) -> Optional[SocketAddress]:
        return self.__local

    #
    #   Threading
    #

    @property
    def running(self) -> bool:
        return self.__running

    async def start(self):
        assert not self.__running, 'server hub is running: %s' % self
        # 1. mark this hub to running
        self.__running = True
        # 2. start an async task for this hub
        self.__daemon.start()
        # await self.run()

    async def stop(self):
        # 1. mark this hub to stopped
        self.__running = False
        # 2. waiting for the hub to stop
        await Runner.sleep(seconds=0.25)
        # 3. cancel the async task
        self.__daemon.stop()

    # Override
    async def run(self):
        self.__running = True
        while self.running:
            master = self._get_master()
            try:
                sock, address = master.accept()
                if sock is None:
                    await Runner.sleep(seconds=Runner.INTERVAL_NORMAL)
                else:
                    await self._accept(remote=address, local=self.local_address, sock=sock)
            except socket.error as error:
                if error.errno == socket.EAGAIN:  # error.strerror == 'Resource temporarily unavailable':
                    if not await is_blocking(sock=master):
                        continue
                print('[TCP] socket error: %s' % error)
            except Exception as error:
                print('[TCP] accept error: %s' % error)

    async def _accept(self, remote: SocketAddress, local: SocketAddress, sock: socket.socket):
        # override for user-customized channel
        channel = self._create_channel(remote=remote, local=local)
        assert isinstance(channel, BaseChannel), 'channel error: %s, %s' % (remote, channel)
        # set socket for this channel
        sock.setblocking(False)
        await channel.set_socket(sock=sock)
        self._set_channel(channel=channel, remote=channel.remote_address, local=channel.local_address)

    # Override
    async def open(self, remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        assert remote is not None, 'remote address empty: %s, %s' % (remote, local)
        return self._get_channel(remote=remote, local=local)


class ClientHub(StreamHub):
    """ Stream Client Hub """

    def __init__(self, delegate: ConnectionDelegate):
        super().__init__(delegate=delegate)
        self.__lock = threading.Lock()

    # Override
    def _create_connection(self, remote: SocketAddress, local: Optional[SocketAddress]) -> Optional[Connection]:
        conn = ActiveConnection(remote=remote, local=local)
        conn.delegate = self.delegate  # gate
        return conn

    # Override
    async def open(self, remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        assert remote is not None, 'remote address empty: %s, %s' % (remote, local)
        # try to get channel
        with self.__lock:
            old = self._get_channel(remote=remote, local=local)
            if old is None:
                # create channel with socket
                channel = self._create_channel(remote=remote, local=local)
                self._set_channel(channel, remote=remote, local=local)
            else:
                channel = old
        if old is None:
            # initialize socket
            sock = await create_socket(remote=remote, local=local)
            if sock is None:
                print('[TCP] failed to prepare socket: %s -> %s' % (local, remote))
                self._remove_channel(channel, remote=remote, local=local)
                channel = None
            else:
                assert isinstance(channel, BaseChannel), 'channel error: %s, %s' % (remote, channel)
                # set socket for this channel
                await channel.set_socket(sock=sock)
        return channel


async def create_socket(remote: SocketAddress, local: Optional[SocketAddress]) -> Optional[socket.socket]:
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 0)
        sock.setblocking(True)
        # try to bind
        if local is not None:
            sock.bind(local)
        # try to connect
        sock.connect(remote)
        sock.setblocking(False)
        return sock
    except socket.error as error:
        print('[TCP] %s > failed to create socket %s -> %s: %s' % (current_time(), local, remote, error))


def current_time() -> str:
    now = time.time()
    localtime = time.localtime(now)
    return time.strftime('%Y-%m-%d %H:%M:%S', localtime)
